design/design-doc-agent.md

```diff
@@ -0,0 +1,627 @@
+connect to the central cluster. Additionally, it will need some permissions to
+execute operations in the remote cluster like `Secret` and
+`CustomResourceDefinition` creation. We will look at how the agent will be
+granted permissions to do its job in two separate domains with different
+mechanisms.
+
+#### Authenticating to The Central Cluster
+
+In order to execute any operation, a `ServiceAccount` needs to exist in the
+central cluster with appropriate permissions to read during pull and write
+during push operations while synchronizing with central cluster. Since the agent
+is running in the remote cluster, the credentials of this `ServiceAccount` will
+be stored in a `Secret` in the remote cluster. Alongside the credentials, the
+agent needs to know the namespace that it should sync to in the central cluster.
+
+The easiest configuration would be the one where we specify the `Secret` and a
+target namespace via installation commands of the agent. However, the namespace
+pairing list could change dynamically after the installation. So, we need to be
+able to supply the pairings after the installation. In order to specify which
+namespace in the remote cluster should be synced to which namespace in the
+central cluster, `Namespace` resource needs to be annotated.
+
+```yaml
+apiVersion: v1
+kind: Namespace
+metadata:
+  name: foo
+  annotations:
+    "agent.crossplane.io/target-namespace": bar
+```
+
+
+We also need to supply which `Secret` in the cluster contains the credentials to
+connect to that target namespace. There are two ways to do that:
+
+* Annotate `Namespace` with the name of the `Secret`.
+  * The `Secret` has to exist in that `Namespace`.
+* If name of the `Secret` is not given as annotation on the `Namespace`, use the
+  `Secret` name supplied as installation configuration.
+
+Here is an example `foo` namespace that uses its own `Secret` to connect to
+target `bar` namespace.
+
+```yaml
+apiVersion: v1
+kind: Namespace
+metadata:
+  name: foo
+  annotations:
+    "agent.crossplane.io/target-namespace": bar
+    "agent.crossplane.io/credentials-secret-name": my-sa-creds
+---
+apiVersion: v1
+kind: Secret
+metadata:
+  name: my-sa-creds
+  namespace: foo
+type: Opaque
+data:
+  kubeconfig: MWYyZDFlMmU2N2Rm...
+```
+
+In case the annotation `agent.crossplane.io/credentials-secret-name` is not
+given on `Namespace`, then the installation configuration will be used which
+would look like the following:
+
+```bash
+helm install --namespace crossplane crossplane/agent --set credentials-secret-name=my-sa-creds
+```
+
+#### Authorization
+
+##### Remote Cluster
+
+Since there will be one agent for the whole cluster, its own mounted
+`ServiceAccount` in that remote cluster needs to have read & write permissions
+for all of the following kinds in the remote cluster listed below:
+
+* `CustomResourceDefinition`
+* `Composition`
+* `InfrastructureDefinition`
+* `InfrastructurePublication`
+* `Secret`
+* All `*Requirement` types
```

Comment:
I assume that the agent in the remote cluster somehow 'read' all those CRDs from the crossplane cluster and then apply them in the remote cluster? or it's something else?

Reply:
The remote cluster is the authority over all fields of the `*Requirement` types except `status`; `status` field will be pulled from Crossplane cluster. The rest of the CRs will be only pulled from the central cluster